code/modules/modular_computers/file_system/programs/generic/ntsl2_interpreter.dm

```diff
@@ -7,93 +7,121 @@
 	requires_ntnet = TRUE
 	available_on_ntnet = TRUE
 
-	nanomodule_path = /datum/nano_module/program/computer_ntsl2_interpreter
-
 	var/datum/ntsl_program/running
+	var/last_terminal_update = 0
 	color = LIGHT_COLOR_GREEN
 
 /datum/computer_file/program/ntsl2_interpreter/process_tick()
 	if(istype(running))
 		running.cycle(30000)
-	..()
+	. = ..()
 
 /datum/computer_file/program/ntsl2_interpreter/kill_program()
-	..()
+	. = ..()
 	if(istype(running))
 		running.kill()
 		running = null
 
+/datum/computer_file/program/ntsl2_interpreter/ui_interact(mob/user)
+	var/datum/vueui/ui = SSvueui.get_open_ui(user, src)
+	if (!ui)
+		ui = new /datum/vueui/modularcomputer(user, src, running ? "mcomputer-generic-ntsl-terminal" : "mcomputer-generic-ntsl-selectfile", 450, 520, "NTSL2+ Interpreter")
+		ui.auto_update_content = TRUE
+	ui.open()
+
+/datum/computer_file/program/ntsl2_interpreter/vueui_transfer(oldobj)
+	for(var/o in SSvueui.transfer_uis(oldobj, src, running ? "mcomputer-generic-ntsl-terminal" : "mcomputer-generic-ntsl-selectfile", 450, 520, "NTSL2+ Interpreter"))
+		var/datum/vueui/ui = o
+		ui.auto_update_content = TRUE
+	return TRUE
+
+/datum/computer_file/program/ntsl2_interpreter/vueui_data_change(var/list/data, var/mob/user, var/datum/vueui/ui)
+	. = ..()
+	data = . || data || list()
+	// Gather data for computer header
+	var/headerdata = get_header_data(data["_PC"])
+	if(headerdata)
+		data["_PC"] = headerdata
+		. = data
+	
+	if(running)
+		if(ui.activeui != "mcomputer-generic-ntsl-terminal")
+			ui.activeui = "mcomputer-generic-ntsl-terminal"
+			spawn(1)
+				. = data
+				ui_interact(user)
+			return .
+		else
+			VUEUI_SET_CHECK(data["running_name"], running.name, ., data)
+			var term = running.get_terminal()
+			
+			// If the terminal has changed, update the time in the program
+			if(data["terminal"] != term)
+				last_terminal_update = world.time
+			
+			// If the last update time has changed in the program, send it to the client.
+			if(data["last_update"] != last_terminal_update)
+				data["terminal"] = term
+				data["last_update"] = last_terminal_update
+				. = data
+	else
+		VUEUI_SET_CHECK(ui.activeui, "mcomputer-generic-ntsl-selectfile", ., data)
+		LAZYINITLIST(data["files"])
+		var/obj/item/computer_hardware/hard_drive/HDD
+
+		if(!computer || !computer.hard_drive)
+			VUEUI_SET_CHECK(ui.activeui, "mcomputer-generic-ntsl-error", ., data)
+			VUEUI_SET_CHECK(data["error"], "I/O ERROR: Unable to access hard drive.", ., data)
+		else
+			HDD = computer.hard_drive
+			for(var/datum/computer_file/F in HDD.stored_files)
```

Comment:
Files get added but never removed. Simple fix is just check if length is diffrent from target, and then reset file list. Still it wouldn't work because of if and it would cause constant updates. 
You should list all files here and add attribute determining if valid and filter on client.
See https://github.com/Aurorastation/Aurora.3/blob/2bdc3cde46d93b06090cfaf5859e553ba34e9a49/code/modules/modular_computers/file_system/programs/generic/records.dm#L135